### PR TITLE
Zoom per zim

### LINKED ISSUE #273: Save zoom level on a ZIM base
1. Default zoom.
    Is there a config file that can control this?
2. Remember zoom (per vim file?)

### PR DESCRIPTION
**Pull request by jetownfeve21**
- add a default zoom field in the settings page
- zoom factor per zim is saved
fix #273

### CODE REVIEW

**Comment by kelson42** [comment]
Unsupported scenario:
* Open a ZIM file
* Change the zoom (with the mouse wheel)
* Open a link in a new tab (mouse wheel double click)
* The new tab has the default zoom

It should have the newly defined zoom.

**Comment by kelson42** [comment]
Zoom is not saved for me over app restart.

**Comment by jetownfeve21** [comment]
> Zoom is not saved for me over app restart.

Even with the keyboard shortcuts ?

**Comment by kelson42** [comment]
> > Zoom is not saved for me over app restart.
> 
> Even with the keyboard shortcuts ?

Yes, it does not work either.

**Comment by jetownfeve21** [comment]
@kelson42 is there a file .config/Kiwix/kiwix-desktop.conf, if yes is there anything in this file ? Something like this : 
[006cbd1b-16d8-b00d-a584-c1ae110a94ed]
zoomFactor=1.3000000000000003

[02e9c7ff-36fc-9c6e-6ac7-cd7085989029]
zoomFactor=1.5000000000000004

[localKiwixServer]
port=8181

[view]
zoomFactor=1

**Comment by kelson42** [comment]
@jetownfeve21 Which file are we talking about? What is the full path?

**Comment by jetownfeve21** [comment]
It is the file where all the settings are saved. According to the Qt documentation its $HOME/.config/Kiwix/kiwix-desktop.conf for Unix
image

**Comment by kelson42** [comment]
```
$ more .config/Kiwix/Kiwix-desktop.conf
[06041133-f108-fe0c-5ff5-3984b188c05e]
zoomFactor=1.8000000000000007

[716e1a7d-f5e2-3646-90e4-a1ae7f37cc4f]
zoomFactor=1.7000000000000006

[localKiwixServer]
port=8080

[view]
zoomFactor=1
```

**Comment by jetownfeve21** [comment]
> Unsupported scenario:
> 
> * Open a ZIM file
> * Change the zoom (with the mouse wheel)
> * Open a link in a new tab (mouse wheel double click)
> * The new tab has the default zoom
> 
> It should have the newly defined zoom.
@kelson42 
I fixed this one, and I add a temporary commit where I print the zoomfactor to see if you get the value on your side

**Comment by kelson42** [comment]
@jetownfeve21 I don't know it that works because I have another (new?) problem. By switching tabs, the zoom seems to often get resed to 100%. Please test all scenarios extensively, and re-ask for review when you are convinced that everything is perfect.

**Comment by jetownfeve21** [comment]
> @jetownfeve21 I don't know it that works because I have another (new?) problem. By switching tabs, the zoom seems to often get resed to 100%. Please test all scenarios extensively, and re-ask for review when you are convinced that everything is perfect.

Your issue is due to the fact that I set the zoomfactor each time the current tab is changed, because it seems to not work on your side it is reset to 100%. Everything work on my side that's why I added a debug commit to know what happens on your side. I maybe have an idea about the cause of the problem but because it works on my side, it's complicated to be sure it's perfect.

**Comment by kelson42** [comment]
@jetownfeve21 Install Ubuntu 18.04 like I have in a VM, I see no fundamental reason why you should not be able to reproduce that problem.

**Comment by jetownfeve21** [comment]
@kelson42 For now it is really difficult for me to set up a new vm because of my connection speed. I have an idea about the cause of the problem, can you test for me my last push and say to me what is displayed on the console, I think it's easier like that for now even if it's not that great

**Comment by kelson42** [comment]
```
$ rm ~/.config/Kiwix/Kiwix-desktop.conf 
$ ./kiwix-desktop 
Library directory : "/home/kelson/.local/share/kiwix"
add widget
zoomFactor :  1
zoomFactor :  1
Handling request QUrl("zim://028055ac-4acc-1d54-65e0-a96de45e1b22.zim/")
Handling request QUrl("zim://028055ac-4acc-1d54-65e0-a96de45e1b22.zim/-/static/bootstrap/css/bootstrap.min.css")
Handling request QUrl("zim://028055ac-4acc-1d54-65e0-a96de45e1b22.zim/-/static/bootstrap/css/bootstrap-theme.min.css")
Handling request QUrl("zim://028055ac-4acc-1d54-65e0-a96de45e1b22.zim/-/static/main.css")
Handling request QUrl("zim://028055ac-4acc-1d54-65e0-a96de45e1b22.zim/-/static/jquery-2.1.4.min.js")
Handling request QUrl("zim://028055ac-4acc-1d54-65e0-a96de45e1b22.zim/-/static/moment.js")
Handling request QUrl("zim://028055ac-4acc-1d54-65e0-a96de45e1b22.zim/I/favicon.png")
Handling request QUrl("zim://028055ac-4acc-1d54-65e0-a96de45e1b22.zim/-/static/main.js")
Handling request QUrl("zim://028055ac-4acc-1d54-65e0-a96de45e1b22.zim/I/static/bootstrap/fonts/glyphicons-halflings-regular.woff2")
session saved
$ more ~/.config/Kiwix/Kiwix-desktop.conf 
[028055ac-4acc-1d54-65e0-a96de45e1b22]
zoomFactor=4.6999999999999975
kelson@rhys:~/code/kiwix-desktop$ ./kiwix-desktop 
Library directory : "/home/kelson/.local/share/kiwix"
add widget
zoomFactor :  1
zoomFactor :  1
Handling request QUrl("zim://028055ac-4acc-1d54-65e0-a96de45e1b22.zim/")
Handling request QUrl("zim://028055ac-4acc-1d54-65e0-a96de45e1b22.zim/-/static/bootstrap/css/bootstrap.min.css")
Handling request QUrl("zim://028055ac-4acc-1d54-65e0-a96de45e1b22.zim/-/static/bootstrap/css/bootstrap-theme.min.css")
Handling request QUrl("zim://028055ac-4acc-1d54-65e0-a96de45e1b22.zim/-/static/main.css")
Handling request QUrl("zim://028055ac-4acc-1d54-65e0-a96de45e1b22.zim/-/static/jquery-2.1.4.min.js")
Handling request QUrl("zim://028055ac-4acc-1d54-65e0-a96de45e1b22.zim/-/static/moment.js")
Handling request QUrl("zim://028055ac-4acc-1d54-65e0-a96de45e1b22.zim/I/favicon.png")
Handling request QUrl("zim://028055ac-4acc-1d54-65e0-a96de45e1b22.zim/-/static/main.js")
Handling request QUrl("zim://028055ac-4acc-1d54-65e0-a96de45e1b22.zim/I/static/bootstrap/fonts/glyphicons-halflings-regular.woff2")
```

**Comment by jetownfeve21** [comment]
@kelson42 Can you test for me again please ?

**Comment by kelson42** [comment]
Still not working:
```
$ more ~/.config/Kiwix/Kiwix-desktop.conf
[006cbd1b-16d8-b00d-a584-c1ae110a94ed]
zoomFactor=2.600000000000001

[028055ac-4acc-1d54-65e0-a96de45e1b22]
zoomFactor=4.6999999999999975$ ./kiwix-desktop 
Library directory : "/home/kelson/.local/share/kiwix"
add widget
zoomFactor :  1
zoomFactor :  1
Handling request QUrl("zim://006cbd1b-16d8-b00d-a584-c1ae110a94ed.zim/")
Handling request QUrl("zim://006cbd1b-16d8-b00d-a584-c1ae110a94ed.zim/-/s/style.css")
Handling request QUrl("zim://006cbd1b-16d8-b00d-a584-c1ae110a94ed.zim/-/s/css_modules/ext.scribunto.logs.css")
Handling request QUrl("zim://006cbd1b-16d8-b00d-a584-c1ae110a94ed.zim/-/j/js_modules/jsConfigVars.js")
Handling request QUrl("zim://006cbd1b-16d8-b00d-a584-c1ae110a94ed.zim/-/j/js_modules/startup.js")
Handling request QUrl("zim://006cbd1b-16d8-b00d-a584-c1ae110a94ed.zim/-/j/js_modules/jquery.js")
Handling request QUrl("zim://006cbd1b-16d8-b00d-a584-c1ae110a94ed.zim/-/j/js_modules/mediawiki.js")
Handling request QUrl("zim://006cbd1b-16d8-b00d-a584-c1ae110a94ed.zim/-/j/js_modules/site.js")
Handling request QUrl("zim://006cbd1b-16d8-b00d-a584-c1ae110a94ed.zim/-/j/js_modules/ext.scribunto.logs.js")
js: Uncaught SyntaxError: Unexpected token <
js: Uncaught SyntaxError: Unexpected token <
js: Uncaught SyntaxError: Unexpected token <
js: Uncaught SyntaxError: Unexpected token <
js: Uncaught SyntaxError: Unexpected token <
^C
```

You have to download an Ubuntu image and test it. Please find a solution to get that working, I don't have the time everyday to test new code/theory for that ticket.

**Comment by jetownfeve21** [comment]
I'm already on Ubuntu 18.04, that's why I don't understand

**Comment by kelson42** [comment]
@jetownfeve21 Install the Flatpak package to secure the dev version and flatpak versions can work properly beside each other.

**Comment by jetownfeve21** [comment]
@kelson42 
I installed the flatpak package and both versions can run beside each other.
I add debug for the config file path, the settings requested, and a message if there is no settings found.

**Comment by kelson42** [comment]
@rgaudin If you could also check if this works fine for you too, that would be helpful. If it works for you then this is probable that my setup is simply somehow broken.

**Comment by rgaudin** [comment]
Works as expected both with mouse and keystrokes on my Virtualbox VM (Ubuntu 18.04)

Here's my conf file.
```more ~/.config/Kiwix/Kiwix-desktop.conf 
[009e8535-22dc-845d-fbda-30dbbd22d058]
zoomFactor=1.7000000000000006

[e983f907-58aa-46d2-2b05-8952fe38a33b]
zoomFactor=1.1
```

File: src/settingsmanager.cpp [hunk 1]
```diff
@@ -18,14 +18,62 @@ SettingsManagerView* SettingsManager::getView()
...
+    qInfo() << "settings file path : " << m_settings.fileName();
+    qInfo() << "Zoom factor id requested : " << id;
+    float zoomFactor = m_zoomFactor;
+    QString key = id + "/zoomFactor";
+    qInfo() << "setting requested : " << key;
+    qInfo() << "list of settings saved : ";
+    for (auto &setting:m_settings.allKeys()) {
+        qInfo() << setting;
+    }
+    if (settingsExists(key)) {
```
Comment (kelson42) [inline src/settingsmanager.cpp]: Please remove empty spaces at end of line

File: resources/texts/_settingsManager.html [hunk 2]
```diff
@@ -13,6 +13,10 @@ <h1>Settings</h1>
         <label>Port for local Kiwix server : </label>
         <input type="number" min="1" max="65535" v-model="kiwixServerPort" v-on:change="setPort()">
     </div>
+    <div class="row">
+        <label>Zoom level : </label>
+        <input type="number" step="10" min="30" max="500" v-model="zoomFactor" v-on:change="setZoomFactor()">
```
Comment (kelson42) [inline resources/texts/_settingsManager.html]: Unity should be given. I guess here this is "%"

File: resources/texts/_settingsManager.html [hunk 3]
```diff
@@ -13,6 +13,10 @@ <h1>Settings</h1>
         <label>Port for local Kiwix server : </label>
```
Comment (kelson42) [inline resources/texts/_settingsManager.html]: This string is not translated/translatable. All the strings should be properly translatable.

File: src/settingsmanager.cpp [hunk 4]
```diff
@@ -18,14 +18,62 @@ SettingsManagerView* SettingsManager::getView()
...
+}
+
+QVariant SettingsManager::getSettings(const QString &key)
+{
+    return m_settings.value(key);
+}
+
+float SettingsManager::getZoomFactorByZimId(const QString &id)
+{
+    qInfo() << "settings file path : " << m_settings.fileName();
```
Comment (kelson42) [inline src/settingsmanager.cpp]: Do you still want so many lines of debug? Do we have a system to handle different level of log messages?

File: resources/texts/_settingsManager.html (hunk 3 above)
Comment (jetownfeve21) [reply resources/texts/_settingsManager.html]: I will fix this in another PR

File: src/settingsmanager.cpp (hunk 4 above)
Comment (jetownfeve21) [reply src/settingsmanager.cpp]: It was temporary debug, I thought I will remove the commit once the problem is fixed. There is no options for different level of log message I think

**Comment by jetownfeve21** [comment]
I changed when the zoom factor is applied because it crashed sometimes, it's fixed now

**Pull request merged**